### PR TITLE
Fixes #222 Added user profile page

### LINKED ISSUE #222: Add a page where user can get auth_token.

### PR DESCRIPTION
**Pull request by sandeepm96**
A simple profile page for user has been added where users detail as well as the auth_token is provided.

Screenshot:
screenshot from 2017-05-03 16 24 00

Suggest changes if needed. :)

### CODE REVIEW

**Comment by adbharadwaj** [comment]
Great start @sandeepm96!!

Here are few comments:
1. Can you explain how a user can reach their profile page ?
2. Can you display the user information using a bootstrap table (will look better) ?
3. Did you confirm that the auth token shown on profile page is correct ? If yes, how ?

**Comment by adbharadwaj** [comment]
Also, can you change the pull request from `master` to `develop` ?

File: templates/user_profile/index.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{% block content %}
+    <div class="container-fluid zero-padding zero-margin">
+
+        <div class="content">
+          <div class="container-fluid">
+
+            <h1>Profile</h1>
+
+              <h5>UserID: {{user.id}}</h5>
+              <h5>Email: {{user.email}}</h5>
```
Comment (adbharadwaj) [inline templates/user_profile/index.html]: Use bootstrap table to make it look better.

File: templates/partials/header.html [hunk 2]
```diff
@@ -36,6 +36,9 @@
                            aria-expanded="false">{{ uid }} <span class="caret"></span></a>
                         <ul class="dropdown-menu">
                             <li class="text-center">
+                                <a href="{% url 'user_profile' %}">Profile <span class="glyphicon glyphicon-user"></span></a>
```
Comment (adbharadwaj) [inline templates/partials/header.html]: Change `Profile` to `My profile`.

File: applications/users/controllers.py [hunk 3]
```diff
@@ -103,6 +104,10 @@ def is_user_authorized_to_update_group(request, username, group_id):
 def get_user(request, email):
 	return db.get_user(request.db_session, email) if email is not None else None
 
+def get_user_profile(request, email):
```
Comment (adbharadwaj) [inline applications/users/controllers.py]: Can we merge `get_user` and `get_user_profile` ? There is only one line difference. Although auth token functionality wouldn't be used by other callers of `get_user` function, there is no need to create a new controller function. What do you think ?

File: applications/users/urls.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 	url(r'^groups/(?P<group_id>[^/]+)$', views.group_page, name='group'),
 
 	url(r'^groups/(?P<group_id>[^/]+)/join/$', views.join_group_page, name='signup_by_invitation'),
+	url(r'^profile/$', views.user_page, name='user_profile'),
```
Comment (adbharadwaj) [inline applications/users/urls.py]: Move to the top. Will be more logical.

File: applications/users/controllers.py (hunk 3 above)
Comment (sandeepm96) [inline applications/users/controllers.py]: `get_user` returns only one value but `get_user_profile` returns two values, i.e. the user object and the auth_token. So their uses will be different from each other. I think keeping this way will be better as we won't have to change everywhere the way `get_user` will be called (accepting two return values). If there could be a better approach, please mention.

File: templates/user_profile/index.html (hunk 1 above)
Comment (sandeepm96) [inline templates/user_profile/index.html]: Yes that would be better.

**Comment by sandeepm96** [comment]
The profile page can be reached from the header dropdown menu beside the user email. As per the suggestions, I am making the required changes and after that, will move the PR to the `develop` branch.

**Comment by sandeepm96** [comment]
The auth_token generated here is the base64 encoding of the user email and the hashed password. But we need the original password for the encoding. Is there anyway we can get the original password for generating the auth_token?

**Comment by sandeepm96** [comment]
The new profile page looks like this:
screenshot from 2017-05-04 00 21 28

File: applications/users/controllers.py (hunk 3 above)
Comment (adbharadwaj) [inline applications/users/controllers.py]: Makes sense. Can you add docstring for `get_user_profile` method ?

File: applications/users/views.py [hunk 5]
```diff
@@ -108,6 +124,27 @@ def join_group_page(request, group_id):
 Users APIs
 '''
 
+def _get_user(request):
```
Comment (adbharadwaj) [inline applications/users/views.py]: We dont need a private function for this functionality. You can directly call the controller function from `user_page` method. 

The reason we had private methods for other operations is that the same code was being repeated in multiple place. There is no need for `_get_user `.

File: templates/user_profile/index.html [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        <div class="content">
+          <div class="container-fluid">
+
+            <h1 class="padding-top-2 padding-bottom-2">My Profile</h1>
+
+            <div class="table-responsive">
+              <table class="table table-bordered table-striped">
+                <tbody>
+                  <tr>
+                    <td><strong>UserID</strong></td>
```
Comment (adbharadwaj) [inline templates/user_profile/index.html]: Change to `User Identifier`

**Comment by sandeepm96** [comment]
I have made the requested changes. I am still in doubt whether generating auth_token from hashed password will not affect the authentication method. Can you clarify regarding this?

**Comment by adbharadwaj** [comment]
fixes #222

**Comment by adbharadwaj** [comment]
@sandeepm96 You are right!! Generating auth_token from hashed password will affect the authentication method. Can you suggest a solution ?

**Comment by sandeepm96** [comment]
We can hash the password and then generate auth_token in the client library instead of directly generating auth_token from the email and password. 
Similarly when a request is made, we will get the email and hashed password from the auth_token. And then we can compare the hashed password with the password that is stored in DB to authorize the user.

**Comment by adbharadwaj** [comment]
Closing this pull request. It doesnt handle the problem generating auth_token from unhashed password.

**Pull request closed without merging**